In this practice code, we are going to use Ozone dataset. Dictionary of this dataset can be found in https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/airquality.html

In [1]:
import pandas as pd
import numpy as np

In [2]:
url = "../../Data/ozone.csv"
OzoneData = pd.read_csv(url)

#### Explore the dataset and decide which variables suffer from missing data

In [3]:
OzoneData.describe()

Ozone     Solar.R        Wind        Temp       Month         Day
count  116.000000  146.000000  153.000000  153.000000  153.000000  153.000000
mean    42.129310  185.931507    9.957516   77.882353    6.993464   15.803922
std     32.987885   90.058422    3.523001    9.465270    1.416522    8.864520
min      1.000000    7.000000    1.700000   56.000000    5.000000    1.000000
25%     18.000000  115.750000    7.400000   72.000000    6.000000    8.000000
50%     31.500000  205.000000    9.700000   79.000000    7.000000   16.000000
75%     63.250000  258.750000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

Answer: Both Ozone and Solar.R are missing data

#### Let's drop rows that have missing values in all the columns you indentified above

Hint: in dropna() if you set how = 'all', it will only drop columns that are suffering from missing values at all varaibles you introduce in subset. If you want to get rid of the row that contains missing values in any of the variables you specify, then you shall set how = 'any'

df.dropna(how = 'all',subset = ['Var1','Var2','Var3'],inplace = True)

The above code will check if all 3 variables specified in df have missing values, if they all have missing values it will drop that row.

In [4]:
OzoneData.dropna(how='all',subset=['Ozone','Solar.R'],inplace=True)

In [5]:
OzoneData.describe()

Ozone     Solar.R        Wind        Temp       Month         Day
count  116.000000  146.000000  151.000000  151.000000  151.000000  151.000000
mean    42.129310  185.931507    9.941722   78.165563    7.019868   15.801325
std     32.987885   90.058422    3.524984    9.198138    1.406984    8.832531
min      1.000000    7.000000    1.700000   57.000000    5.000000    1.000000
25%     18.000000  115.750000    7.400000   73.000000    6.000000    8.000000
50%     31.500000  205.000000    9.700000   79.000000    7.000000   16.000000
75%     63.250000  258.750000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

#### Spoiler! If everything is going according to plan you shall be left by 151 observations. Also, it seemed like the first two varibales had missing values. Now plese make a copy of your dataframe into a dataframe named OzoneImputeMean. Also, please use mean of the variables to fill in missing values in OzoneImputeMean

In [6]:
OzoneImputeMean = OzoneData.copy()

In [7]:
#Check if missing values are filled

OzoneImputeMean['Ozone'].fillna(value = np.mean(OzoneImputeMean['Ozone']), inplace = True)


In [8]:
OzoneData.Month.unique()

array([5, 6, 7, 8, 9])

#### Now it's time for imputing using linear regression lines

In [9]:
#### Before we start let's define dummy variables for variable Month - don't worry about day!
MonthDummy = pd.get_dummies(OzoneData.Month, prefix = 'Month')
del MonthDummy['Month_9']
OzoneData = pd.concat([OzoneData, MonthDummy], axis=1)
OzoneData.head()

Ozone  Solar.R  Wind  Temp  Month  Day  Month_5  Month_6  Month_7  Month_8
0     41      190   7.4    67      5    1        1        0        0        0
1     36      118   8.0    72      5    2        1        0        0        0
2     12      149  12.6    74      5    3        1        0        0        0
3     18      313  11.5    62      5    4        1        0        0        0
5     28      NaN  14.9    66      5    6        1        0        0        0

In [10]:
#now let's explore correlation Matrix
OzoneData.corr()

Ozone   Solar.R      Wind      Temp     Month       Day   Month_5  \
Ozone    1.000000  0.348342 -0.601547  0.698360  0.164519 -0.013226 -0.302963   
Solar.R  0.348342  1.000000 -0.056792  0.275840 -0.075301 -0.150275 -0.024601   
Wind    -0.601547 -0.056792  1.000000 -0.466032 -0.175317  0.042365  0.237781   
Temp     0.698360  0.275840 -0.466032  1.000000  0.397427 -0.136876 -0.637816   
Month    0.164519 -0.075301 -0.175317  0.397427  1.000000 -0.007727 -0.702257   
Day     -0.013226 -0.150275  0.042365 -0.136876 -0.007727  1.000000  0.011003   
Month_5 -0.302963 -0.024601  0.237781 -0.637816 -0.702257  0.011003  1.000000   
Month_6 -0.112006  0.023998  0.046054  0.050753 -0.362131 -0.017044 -0.242766   
Month_7  0.277961  0.176744 -0.144638  0.318103 -0.007201  0.011471 -0.247805   
Month_8  0.291808 -0.076390 -0.166105  0.321680  0.355246  0.011471 -0.247805   

          Month_6   Month_7   Month_8  
Ozone   -0.112006  0.277961  0.291808  
Solar.R  0.023998  0.176744 -0.076390  
Wind     0.046054 -0.144638 -0.166105  
Temp     0.050753  0.318103  0.321680  
Month   -0.362131 -0.007201  0.355246  
Day     -0.017044  0.011471  0.011471  
Month_5 -0.242766 -0.247805 -0.247805  
Month_6  1.000000 -0.253080 -0.253080  
Month_7 -0.253080  1.000000 -0.258333  
Month_8 -0.253080 -0.258333  1.000000

#### What seems to be the list of best variables can define Ozone? how about Solar.R?

Temp x ozone are positively correlated, and Wind x Ozone are negatively correlated.

Solar radiation is also weekly correlated with ozone. 

Months 7 and 8 (July/Aug) are correlated with temp (no surprise there)

In [11]:
# now let's use a regression model to predict Ozone. 
# First drop NaN values in Ozone and save it in OzoneDroppedValues_Ozone
# the run a regression line on variables of interest and check significancy of your model
# if in a multi class dummy variable case, you see a few of the dummy variables are not significant but the rest are, 
#you shall either drop all or keep all. Otherwise, selecting the base dummy will become important
# use these variables ['Solar.R','Wind','Temp','Month_5','Month_6','Month_7','Month_8'] to predict
from sklearn.linear_model import LinearRegression
from sklearn import feature_selection


In [12]:
OzoneDroppedValues_Ozone = OzoneData.dropna(how='any',subset=['Ozone','Solar.R'])

In [13]:
OzoneDroppedValues_Ozone.describe()

Ozone     Solar.R        Wind        Temp       Month         Day  \
count  111.000000  111.000000  111.000000  111.000000  111.000000  111.000000   
mean    42.099099  184.801802    9.939640   77.792793    7.216216   15.945946   
std     33.275969   91.152302    3.557713    9.529969    1.473434    8.707194   
min      1.000000    7.000000    2.300000   57.000000    5.000000    1.000000   
25%     18.000000  113.500000    7.400000   71.000000    6.000000    9.000000   
50%     31.000000  207.000000    9.700000   79.000000    7.000000   16.000000   
75%     62.000000  255.500000   11.500000   84.500000    9.000000   22.500000   
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000   

          Month_5     Month_6     Month_7     Month_8  
count  111.000000  111.000000  111.000000  111.000000  
mean     0.216216    0.081081    0.234234    0.207207  
std      0.413530    0.274198    0.425440    0.407143  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000    0.000000  
75%      0.000000    0.000000    0.000000    0.000000  
max      1.000000    1.000000    1.000000    1.000000

In [14]:
linreg = LinearRegression()
X = OzoneDroppedValues_Ozone[['Solar.R','Wind','Temp','Month_5','Month_6','Month_7','Month_8']]
Y = OzoneDroppedValues_Ozone['Ozone']
linreg.fit(X,Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [15]:
# now fill in null values of OzoneData['Ozone'] by predicted values
OzoneDroppedValues_Ozone['Ozone_p'] = linreg.predict(X)

/Users/albarron/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app


In [16]:
# Now repeat previous steps for Solar.R variable using ['Ozone','Wind','Temp]

X = OzoneDroppedValues_Ozone[['Ozone','Wind','Temp']]
Y = OzoneDroppedValues_Ozone['Solar.R']
linreg.fit(X,Y)
# OzoneDroppedValues_Ozone['Solar.R_p'] = linreg.predict(X)
# OzoneData['Solar.R'].fillna(OzoneDroppedValues_Ozone['Solar.R_p'], inplace = True)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [17]:
#Check p-values of your model - if a variable is not significant, drop it
feature_selection.f_regression(X,Y)[1]
# Ozone: .000179310857
# Wind : .183451976
# Temp : .00173078565


array([  1.79310857e-04,   1.83451976e-01,   1.73078565e-03])

In [18]:
### We shall drop Wind and rerun our regressions again
X = OzoneDroppedValues_Ozone[['Ozone','Temp']]
Y = OzoneDroppedValues_Ozone['Solar.R']
linreg.fit(X,Y)
feature_selection.f_regression(X,Y)[1]
# Ozone: .00017931
# Temp : .00173079
# The pvalues are effectively the same?

array([ 0.00017931,  0.00173079])

In [19]:
# now fill in null values of OzoneData['Solar.R'] by predicted values
OzoneDroppedValues_Ozone['Solar.R_p'] = linreg.predict(X)

/Users/albarron/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app


#### Now check your filled data - if your predicted values are more than maximum or less than minimum, replace them by max and min

In [20]:
# This is one way to do it for Ozone
OzoneDroppedValues_Ozone.loc[(OzoneDroppedValues_Ozone['Ozone_p'] >= OzoneData['Ozone'].max()),'Ozone_p'].apply(lambda x:OzoneData['Ozone'].max())

Series([], Name: Ozone_p, dtype: float64)

In [47]:
# The other way
y_max = OzoneData['Solar.R'].max()
OzoneDroppedValues_Ozone['Solar.R_p'] = [y_max if x > y_max else x for x in OzoneDroppedValues_Ozone['Solar.R_p']]

/Users/albarron/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


In [48]:
OzoneData['Solar.R'].fillna(OzoneDroppedValues_Ozone['Solar.R_p'], inplace = True)
OzoneData['Ozone'].fillna(OzoneDroppedValues_Ozone['Ozone_p'], inplace = True)

## Bonus: In original question, replace missing values of Ozone and Solar.R by Single Imputation Regression Line with Error

In [60]:
url = "../../Data/ozone.csv"
OzoneData = pd.read_csv(url)
OzoneData_dropped = OzoneData.dropna(how = 'all', subset = ['Ozone', 'Solar.R'])
#Set baseline
OzoneData.describe()

Ozone     Solar.R        Wind        Temp       Month         Day
count  116.000000  146.000000  151.000000  151.000000  151.000000  151.000000
mean    42.129310  185.931507    9.941722   78.165563    7.019868   15.801325
std     32.987885   90.058422    3.524984    9.198138    1.406984    8.832531
min      1.000000    7.000000    1.700000   57.000000    5.000000    1.000000
25%     18.000000  115.750000    7.400000   73.000000    6.000000    8.000000
50%     31.500000  205.000000    9.700000   79.000000    7.000000   16.000000
75%     63.250000  258.750000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

In [58]:

# Ozone fill
X = OzoneData_dropped[['Solar.R','Wind','Temp']]
Y = OzoneData_dropped['Ozone']
y_hat = linreg.predict(X)
stderr = (sum((Y-y_hat)**2)/(len(Y)-2))**.5
ozone_y_hat_error = y_hat + np.random.normal(0,stderr,len(OzoneData_dropped['Ozone']))



# Solar.R fill
ozone_y_hat_error = y_hat + np.random.normal(0,stderr,len(OzoneData_dropped['Solar.R']))
X = OzoneData_dropped[['Ozone','Temp']]
Y = OzoneData_dropped['Solar.R']
y_hat = linreg.predict(X)
stderr = (sum((Y-y_hat)**2)/(len(Y)-2))**.5
solarr_y_hat_error = y_hat + np.random.normal(0,stderr,len(OzoneData_dropped['Solar.R']))



ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

In [51]:
OzoneData['Solar.R'].fillna(pd.Series(ozone_y_hat_error),inplace = True)
OzoneData['Ozone'].fillna(pd.Series(solarr_y_hat_error),inplace = True)


In [53]:
OzoneData.describe()

Ozone     Solar.R        Wind        Temp       Month         Day
count  116.000000  153.000000  153.000000  153.000000  153.000000  153.000000
mean    42.129310  183.989781    9.957516   77.882353    6.993464   15.803922
std     32.987885   88.906060    3.523001    9.465270    1.416522    8.864520
min      1.000000    7.000000    1.700000   56.000000    5.000000    1.000000
25%     18.000000  113.144703    7.400000   72.000000    6.000000    8.000000
50%     31.500000  197.000000    9.700000   79.000000    7.000000   16.000000
75%     63.250000  256.000000   11.500000   85.000000    8.000000   23.000000
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000

## 